src/logging.cpp

```diff
@@ -224,10 +224,32 @@ std::string BCLog::Logger::LogTimestampStr(const std::string& str)
     return strStamped;
 }
 
+namespace {
+    /** Belts and suspenders: make sure outgoing log messages don't contain
+     * potentially suspicious characters, such as terminal control codes.
+     *
+     * This escapes control characters except newline ('\n') in C syntax.
+     * It escapes instead of removes them to still allow for troubleshooting
+     * issues where they accidentally end up in strings.
+     */
+    std::string LogEscapeMessage(const std::string& str) {
+        std::string ret;
+        for (char ch_in: str) {
+            uint8_t ch = (uint8_t)ch_in;
+            if (ch >= 32 || ch == '\n') {
+                ret += ch_in;
+            } else {
+                ret += strprintf("\\x%02x", ch);
```

Comment:
Could add a `DCHECK(false)` (#16136) to the else branch, maybe only run once?

Reply:
I think crashing the executable on an assertion error is a bad idea here. After all, if the character was injected somehow, that changes a log injection to a live DoS.

Reply:
Only running the else branch once is fine with me though, not sure. It's a compromise between precise troubleshooting and the fear that this would generate very long strings.